subprojects/tooling-api/src/main/java/org/gradle/tooling/UnsupportedVersionException.java

```diff
@@ -15,17 +15,60 @@
  */
 package org.gradle.tooling;
 
+import org.gradle.util.DeprecationLogger;
+import org.gradle.util.GradleVersion;
+
 /**
  * Thrown when the target Gradle version does not support a particular feature.
  *
  * @since 1.0-milestone-3
  */
 public class UnsupportedVersionException extends GradleConnectionException {
+    private static final String UNSUPPORTED_MESSAGE = "Support for %s older than %s has been removed.%s You should upgrade your %s to version %s or later.";
+    private static final String DEPRECATION_WARNING = "Support for %s older than %s is deprecated and will be removed in 5.0.%s You should upgrade your %s.";
+
+    public static final GradleVersion MIN_PROVIDER_VERSION = GradleVersion.version("1.2");
+
+    public static final GradleVersion MIN_LTS_PROVIDER_VERSION = GradleVersion.version("2.6");
+
     public UnsupportedVersionException(String message) {
         super(message);
     }
 
     public UnsupportedVersionException(String message, Throwable cause) {
         super(message, cause);
     }
+
+    public static UnsupportedVersionException unsupportedProvider() {
```

Comment:
❌ This is a public type, we should keep the API surface minimal. Instead, this logic belongs into the class that does the actual deprecation check.

Reply:
Thanks for reminding. I've reverted `UnsupportedVersionException` and created a new class `ToolingApiVersionHelper` in `org.gradle.tooling.internal` to handle all version related stuff.